### PR TITLE
Git loader + Antlr in the fly parser

### PR DESCRIPTION
**Pull request by raulraja**
```kotlin
package com.xebia.functional.xef.auto.git

import arrow.core.raise.recover
import com.xebia.functional.xef.antlr.languages.Language
import com.xebia.functional.xef.antlr.languages.antlrParser
import com.xebia.functional.xef.auto.ai
import com.xebia.functional.xef.auto.getOrThrow
import com.xebia.functional.xef.auto.promptMessage
import com.xebia.functional.xef.loaders.FileContent
import com.xebia.functional.xef.loaders.git

suspend fun main() {
  ai {
    contextScope(arrowRepositoryDocuments()) {
      while (true) {
        print("Enter your question: ")
        val line = readlnOrNull() ?: break
        val response = promptMessage(line)
        println(response.firstOrNull())
      }
    }
  }.getOrThrow()
}

private suspend fun arrowRepositoryDocuments() =
  git(
    url = "https://github.com/arrow-kt/arrow",
    fileLoader = ::parseFileWithAntlrIfPossible,
    fileFilter = { it.name.endsWith(".kt") }
  )

private suspend fun parseFileWithAntlrIfPossible(fileContent: FileContent): List<String> {
  val langParser = recover({
    Language.loadByPath(fileContent.path)?.let { language -> antlrParser(language) }
  }) {
    error(it.reason)
  }
  return if (langParser != null) {
    val (_, context, docs) = langParser(fileContent.content)
    if (docs.isNotEmpty()) println("found ${docs.size} docs in ${context::class.java.name}")
    val foundDocuments = docs.map { doc ->
      doc.toDocument(fileContent.path)
    }
    foundDocuments
  } else listOf(fileContent.toDocument())
}
```

The Kotlin parser is defined as:

```kotlin
package com.xebia.functional.xef.antlr.languages

import arrow.core.raise.Raise
import com.xebia.functional.xef.antlr.ANTLRError
import com.xebia.functional.xef.antlr.EntryPoint
import com.xebia.functional.xef.antlr.G4ParserFromURLs

fun Raise<ANTLRError>.kotlin() =
  G4ParserFromURLs(
    EntryPoint(
      method = "kotlinFile",
      lexer = "KotlinLexer",
      parser = "KotlinParser",
      contextDocuments = setOf("KotlinFile", "Declaration"),
    ),
    "https://raw.githubusercontent.com/antlr/grammars-v4/master/kotlin/kotlin/UnicodeClasses.g4",
    "https://raw.githubusercontent.com/antlr/grammars-v4/master/kotlin/kotlin/KotlinLexer.g4",
    "https://raw.githubusercontent.com/antlr/grammars-v4/master/kotlin/kotlin/KotlinParser.g4",
  )
```

This enables a user to dynamically parse sources in repositories for the languages they care about.
We can extend the pattern to add some heuristics and dynamically fetch and create a parser for any existing language using data sources such as https://github.com/antlr/grammars-v4

### CODE REVIEW

File: core/src/commonMain/kotlin/com/xebia/functional/xef/auto/DeserializerLLMAgent.kt [hunk 1]
```diff
@@ -39,7 +39,7 @@ suspend inline fun <reified A> AIScope.prompt(
   user: String = "testing",
   echo: Boolean = false,
   n: Int = 1,
-  temperature: Double = 0.0,
+  temperature: Double = 0.5,
```
Comment (serras) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/auto/DeserializerLLMAgent.kt:42]: Why is this change part of this PR?

File: examples/kotlin/build.gradle.kts [hunk 2]
```diff
@@ -23,11 +23,14 @@ dependencies {
...
+  implementation(projects.xefGit)
+  implementation(projects.xefAntlr)
   implementation(libs.kotlinx.serialization.json)
   implementation(libs.logback)
   implementation(libs.klogging)
   implementation(libs.bundles.arrow)
   implementation(libs.okio)
   implementation(libs.jdbc.mysql.connector)
   api(libs.bundles.ktor.client)
+  implementation("org.antlr:antlr4-runtime:4.13.0")
```
Comment (serras) [inline examples/kotlin/build.gradle.kts:35]: This should go to the TOML file, I guess.

File: filesystem/src/commonMain/kotlin/com/xebia/functional/xef/io/FileUtils.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  source.buffer().use { src ->
+    while (src.read(buffer).also { bytesRead = it } != -1 && !channel.isClosedForWrite) {
+      channel.writeFully(buffer, offset = 0, length = bytesRead)
+      channel.flush()
+    }
+  }
+}
+
+suspend fun readUrlToSink(url: String, sink: Sink) {
+  HttpClient {  }.use {
```
Comment (serras) [inline filesystem/src/commonMain/kotlin/com/xebia/functional/xef/io/FileUtils.kt:38]: Note that this is not checking the result status code. If you have a 404, for example, this will give you the body with the error message, which may not be what you want.

```suggestion
  HttpClient { expectSuccess = true }.use {
```

File: filesystem/src/jsMain/kotlin/com/xebia/functional/xef/io/NodeCommandExecutor.kt [hunk 4]
```diff
@@ -15,6 +15,34 @@ actual val CommandExecutor.Companion.DEFAULT: CommandExecutor
   get() = NodeCommandExecutor
 
 object NodeCommandExecutor : CommandExecutor {
+
+  override suspend fun platform(): Platform {
```
Comment (serras) [inline filesystem/src/jsMain/kotlin/com/xebia/functional/xef/io/NodeCommandExecutor.kt:19]: I have two questions regarding this:
1. Why is this needed? At the end of the day this always returns `LINUX(something)`, so we could just set this to `LINUX("node")` and call it a day.
2. The implementation seems very brittle. There are modules in Node which do this, and I think delegating on those would be the right move (if getting the platform name is really required).

Comment (serras) [reply filesystem/src/jsMain/kotlin/com/xebia/functional/xef/io/NodeCommandExecutor.kt:19]: Another possibility is to use the [Platform type](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.native/-platform/) available in the JS target.

**Review by serras — CHANGES_REQUESTED** [review]

File: core/src/commonMain/kotlin/com/xebia/functional/xef/auto/DeserializerLLMAgent.kt (hunk 1 above)
Comment (raulraja) [reply core/src/commonMain/kotlin/com/xebia/functional/xef/auto/DeserializerLLMAgent.kt:42]: outdated changes, this branch needs to be merged with `main`. I'll update it.

**Comment by raulraja** [comment]
I'm not sure If I'll have time to update this today, but I'll get to it after the 21st when I get back from vacation. Thanks @serras  for looking into it.

**Comment by raulraja** [comment]
I am closing this to focus on other priorities. In the last couple of months, I learned this is not the best approach, and I plan on rewriting this feature when I have time based on the ReAct agent and a set of git tools.
